lxd/ubuntu-pro/watcher.go

```diff
@@ -0,0 +1,252 @@
+package pro
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/logger"
+)
+
+// GuestAttachSetting represents ubuntu-pro settings for guest attachment availability.
+type GuestAttachSetting string
+
+const (
+	// GuestAttachSettingOff indicates that guest attachment is never available. The value of `GET /1.0/ubuntu-pro` over
+	// devlxd should always be "off".
+	GuestAttachSettingOff GuestAttachSetting = "off"
+
+	// GuestAttachSettingAvailable indicates that guest attachment is available and can be opted into via guest config.
+	// The value of `GET /1.0/ubuntu-pro` over devlxd should return the value of the `ubuntu_pro.guest_attach`
+	// configuration of the instance.
+	GuestAttachSettingAvailable GuestAttachSetting = "available"
+
+	// GuestAttachSettingOn indicates that guest attachment is on and can be opted into via guest config. The value of
+	// `GET /1.0/ubuntu-pro` over devlxd should return the value of the `ubuntu_pro.guest_attach` configuration of the
+	// instance.
+	//
+	// NOTE: This setting currently behaves in the same way as GuestAttachSettingAvailable, but we will use this setting
+	// to decide whether automatically set the value of `ubuntu_pro.guest_attach` on instances based on the origin of
+	// their base image when an authenticated simplestreams remote for pro images becomes available.
+	GuestAttachSettingOn GuestAttachSetting = "on"
+)
+
+// isValid returns an error if the GuestAttachSetting is not one of the pre-defined values.
+func (g GuestAttachSetting) isValid() error {
+	if !shared.ValueInSlice(g, []GuestAttachSetting{GuestAttachSettingOff, GuestAttachSettingAvailable, GuestAttachSettingOn}) {
+		return fmt.Errorf("Invalid guest auto-attach setting %q", g)
+	}
+
+	return nil
+}
+
+// ubuntuAdvantageDirectory is the base directory for ubuntu-pro related configuration.
+const ubuntuAdvantageDirectory = "/var/lib/ubuntu-advantage"
+
+// interfacesDir returns the path to the interfaces directory based on the given ubuntu advantage directory.
+func interfacesDir(ubuntuAdvantageDir string) string {
+	return filepath.Join(ubuntuAdvantageDir, "interfaces")
+}
+
+// lxdConfigFile returns the path to the ubuntu-pro LXD configuration file based on the given "interfaces" directory.
+func lxdConfigFile(interfacesDir string) string {
+	return filepath.Join(interfacesDir, "lxd-config.json")
+}
+
+// Settings contains methods for returning ubuntu-pro related settings.
+type Settings interface {
```

Comment:
Having an interface here feels a bit unnecessary to me.  There's only one implementation.
And in general interfaces are better defined where they are consumed rather than where the implementations are defined.

https://go.dev/wiki/CodeReviewComments

> Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values. The implementing package should return concrete (usually pointer or struct) types: that way, new methods can be added to implementations without requiring extensive refactoring.